Foundation/FileManager+Win32.swift

```diff
@@ -299,9 +299,58 @@ extension FileManager {
     }
 
     internal func _destinationOfSymbolicLink(atPath path: String) throws -> String {
-        return try _canonicalizedPath(toFileAtPath: path)
+        let handle = path.withCString(encodedAs: UTF16.self) { symlink in
+            CreateFileW(symlink,
+                GENERIC_READ,
+                DWORD(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE),
+                nil,
+                DWORD(OPEN_EXISTING),
+                DWORD(FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS),
+                nil)
+        }
+
+        guard handle != INVALID_HANDLE_VALUE else {
+          throw _NSErrorWithWindowsError(GetLastError(), reading: true)
+        }
+        defer { CloseHandle(handle) }
+
+        // Since REPARSE_DATA_BUFFER ends with an arbitrarily long buffer, we
+        // have to manually get the path buffer out of it since binding it to a
+        // type will truncate the path buffer
+        let pathBufferOffset =
+            MemoryLayout<ULONG>.size //ULONG ReparseTag
+            + MemoryLayout<USHORT>.size //ULONG ReparseDataLength
+            + MemoryLayout<USHORT>.size //ULONG Reserved
+            + MemoryLayout<USHORT>.size //ULONG SubstituteNameOffset
+            + MemoryLayout<USHORT>.size //ULONG SubstituteNameLength
+            + MemoryLayout<USHORT>.size //ULONG PrintNameOffset
+            + MemoryLayout<USHORT>.size //ULONG PrintNameLength
+            + MemoryLayout<ULONG>.size //ULONG Flags
+
+        let buffSize = MemoryLayout<REPARSE_DATA_BUFFER>.size + Int(2 * MAX_PATH) * MemoryLayout<wchar_t>.size
+        let buff = UnsafeMutableRawBufferPointer.allocate(byteCount: buffSize, alignment: 8)
+
+        var bytesWritten: DWORD = 0
+        guard DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, nil, 0,
+                buff.baseAddress, DWORD(buffSize), &bytesWritten, nil) else {
+            throw _NSErrorWithWindowsError(GetLastError(), reading: true)
+        }
+
+        let bound = buff.bindMemory(to: REPARSE_DATA_BUFFER.self)
+        guard let reparseDataBuffer = bound.first else {
+            fatalError()
+        }
+
+        let pathBufferPtr = buff.baseAddress! + pathBufferOffset
+        let printNameBytes = Int(reparseDataBuffer.SymbolicLinkReparseBuffer.PrintNameLength)
+        let printNameOffset = Int(reparseDataBuffer.SymbolicLinkReparseBuffer.PrintNameOffset)
```

Comment:
1. This is assuming that the reparse point is a symbolic link. Not all reparse points are symbolic links, or even links to files at all. Please verify the reparse tag before doing anything here. If you want an example of a reparse point which is not a link, connect a OneDrive account on Windows 10 and look in your local OneDrive directory. The files there all use a private reparse tag. (Also, do we want to handle mount points here? They can be links to directories.)

2. Why is this using the print name? The print name is only intended to be used for display purposes.

3. Do we need to validate that the name offset and length actually make sense? We shouldn't assume that reparse data is well-formed. AFAIK, UnsafeMutableRawBufferPointer is not always bounds-checked.